File: collection/list/array.js
function byLowerCase(a, b) {
  if (a.toLowerCase() < b.toLowerCase()) return -1;
  if (a.toLowerCase() > b.toLowerCase()) return 1;
  return 0;
}
function byLength(a, b) {
  if (a.length < b.length) return -1;
  if (a.length > b.length) return 1;
  return 0;
}

let a = ['Art', 'b', 'ART'];

// 1. Creating a new sorted array with a comparator
let b = [...a];

b.sort((a, b) => byLowerCase(a, b));
console.log(b);

// 2. Creating a new sorted array with multiple comparators
// Since 0 value is evaluated to false, the OR hack works nicely here
let c = [...a];

c.sort((a, b) => byLength(a, b) || byLowerCase(a, b));
console.log(c);

// 3. Sorting array with in-place sort
a.sort((a, b) => byLowerCase(a, b));
console.log(a);

// 4. Sorting numeric array to string order
let d = [2, 100, 3];
d.sort((a, b) => {
  if (a.toString() < b.toString()) return -1;
  if (a.toString() > b.toString()) return 1;
  return 0;
});
console.log(d);

// 5. Sorting string array to numeric order
var e = ['2', '100', '3'];
e.sort((a, b) => {
  if (Number(a) < Number(b)) return -1;
  if (Number(a) > Number(b)) return 1;
  return 0;
});
console.log(e);

File: collection/list/array.go
package main

import (
	"fmt"
	"sort"
	"strings"
)

type byLength []string

func (s byLength) Len() int {
	return len(s)
}

func (s byLength) Swap(i, j int) {
	s[i], s[j] = s[j], s[i]
}

func (s byLength) Less(i, j int) bool {
	return len(s[i]) < len(s[j])
}

type byLowerCase []string

func (s byLowerCase) Len() int {
	return len(s)
}

func (s byLowerCase) Swap(i, j int) {
	s[i], s[j] = s[j], s[i]
}

func (s byLowerCase) Less(i, j int) bool {
	return strings.ToLower(s[i]) < strings.ToLower(s[j])
}

type byLengthThenLowerCase []string

func (s byLengthThenLowerCase) Len() int {
	return len(s)
}

func (s byLengthThenLowerCase) Swap(i, j int) {
	s[i], s[j] = s[j], s[i]
}

func (s byLengthThenLowerCase) Less(i, j int) bool {
	if len(s[i]) < len(s[j]) {
		return true
	}
	if len(s[i]) > len(s[j]) {
		return false
	}
	return strings.ToLower(s[i]) < strings.ToLower(s[j])
}

func main() {

	a := []string{"Art", "b", "ART"}
	c := make([]string, len(a))
	copy(c, a)

	// 1. Creating a new sorted array with a comparator
	sort.Sort(byLowerCase(a))
	sort.Sort(byLength(a))
	sort.Sort(byLengthThenLowerCase(a))
	fmt.Println(a)
	fmt.Println(c)
	// sort.Slice(a, func(i, j int) bool {
	// 	return len(a[i]) < len(a[j])
	// })
	x := []int{1, 2, 3, 4}
	fmt.Println(x[2:4])
}
